File: java/fibonacciNumber.java
class Solution {
    private Map<Integer, Integer> memo = new HashMap<>();
    
    public int fib(int n) {
        //recursive base cases
        if (n == 0) {
            return 0;
        }
        if (n == 1) {
            return 1;
        }
        //check if we've already solved this problem
        if (memo.keySet().contains(n)) {
            return memo.get(n);
        }
        //get result recursively and add it to memo
        int result = fib(n - 1) + fib(n - 2);
        memo.put(n, result);
        return result;
    }
}

File: python/fibonacci_number.py
class Solution:
    #store recursive base cases in the cache
    memo = { 0: 0, 1: 1 }
    def fib(self, n: int) -> int:
        #check if problem is already in cache
        if n in self.memo:
            return self.memo[n]
        #otherwise get solution recursively and store it in cache
        result = self.fib(n - 1) + self.fib(n - 2)
        self.memo[n] = result
        return result